internal/monitors/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,276 @@
+	// Enable enhanced index level index stats groups. A list of index stats groups
+	// for which to collect enhanced stats
+	EnableEnhancedIndexStatsForIndexGroups []string `yaml:"enableEnhancedIndexStatsForIndexGroups"`
+	// To enable index stats from only primary shards. By default the index stats collected
+	// are aggregated across all shards
+	EnableIndexStatsPrimaries bool `yaml:"enableIndexStatsPrimaries"`
+}
+
+// Monitor for conviva metrics
+type Monitor struct {
+	Output  types.Output
+	cancel  context.CancelFunc
+	ctx     context.Context
+	timeout time.Duration
+}
+
+func init() {
+	monitors.Register(monitorType, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+// Configure monitor
+func (m *Monitor) Configure(conf *Config) error {
+	esClient := client.NewESClient(conf.Host, conf.Port, conf.UseHTTPS, conf.Username, conf.Password)
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	var nodeIsCurrentMaster bool
+	var defaultDimensions map[string]string
+
+	// Collect node stats
+	utils.RunOnInterval(m.ctx, func() {
+		// Collect info about master for the cluster
+		masterInfoOutput, err := esClient.GetMasterNodeInfo()
+
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to GET master node info")
+			return
+		}
+
+		// Collect Node level stats
+		nodeStatsOutput, err := esClient.GetNodeAndThreadPoolStats()
+
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to GET node stats")
+			return
+		}
+
+		nodeDimensions, err := prepareNodeMetricsDimensions(nodeStatsOutput.NodeStats)
+
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to prepare node dimensions")
+			return
+		}
+
+		nodeIsCurrentMaster, err = isCurrentMaster(nodeDimensions["node_id"], masterInfoOutput)
+
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to determine whether current node is the master")
+			return
+		}
+
+		clusterName := nodeStatsOutput.ClusterName
+		defaultDimensions, err = prepareDefaultDimensions(conf.Cluster, clusterName)
+
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to prepare plugin_instance dimension")
+			return
+		}
+
+		nodeMetricsDims := utils.MergeStringMaps(defaultDimensions, nodeDimensions)
+
+		dps := client.GetNodeStatsDatapoints(nodeStatsOutput, nodeMetricsDims, utils.StringSliceToMap(conf.ThreadPools), utils.StringSliceToMap(conf.EnableEnhancedNodeStatsForIndexGroups), map[string]bool{
+			client.HTTPStatsGroup:       conf.EnableEnhancedHTTPStats,
+			client.JVMStatsGroup:        conf.EnableEnhancedJVMStats,
+			client.ProcessStatsGroup:    conf.EnableEnhancedProcessStats,
+			client.ThreadpoolStatsGroup: conf.EnableEnhancedThreadPoolStats,
+			client.TransportStatsGroup:  conf.EnableEnhancedTransportStats,
+		})
+
+		// If current node is not the master, collect only node level stats
+		if !nodeIsCurrentMaster {
+			m.sendDatapoints(dps)
+			return
+		}
+
+		// Collect cluster level stats
+		clusterStatsOutput, err := esClient.GetClusterStats()
+
+		if err != nil {
```

Comment:
Send the datapoints as they are returned by these methods called in this method.  Don't wait until the end to send all of them.  In the case of this error, you wouldn't even send the node stats if you couldn't fetch the cluster stats at all since you just return here, but also you would delay the sending of the node stats unnecessarily until you got the cluster stats.

Also, parallelize as many of these requests as possible to reduce the runtime of this method and to try and keep them as consistent as possible on the interval.

Reply:
Sending datapoints as soon as they are collected now. I've updated so that fetching cluster stats and index stats can be done in parallel once the information required by both the groups of stats are collected. Please take a look and let me know what you think.